README.md

```diff
@@ -37,3 +37,134 @@ and JavaScript/TypeScript. In order to set up the pre-commit hooks in the JS/TS
 the `bootstrap.js` script was added. With `pre-commit` installed, running the `bootstrap.js` script is equivalent
 to `husky install`. Note that the hooks are only activated when the package is installed locally (`npm install`)
 and not when installed as a dependency.
+
+## Writing tests
+
+### Common architecture
+
+![architecture](image.png)
+
+1. **Feature** is the test case. We use Cucumbers Gherkin syntax to define the test case structure. Each feature consists of one or more steps.
+2. **Step** one particular step for a user to do. It can also describe some background processes (like the creation of workflows). Uses widgets to define what exactly needs to be done in the step.
+3. **Widget** - Javascript class that encapsulates the logic of one particular element on the page (button/dropdown/sidebar/header or even a whole page)
+4. **Browser** - Intermediate level between Widgets and Cypress. Aimed to encapsulate/hide the test engine (cypress or chimp). This is the only level that uses cypress.
+5. **Cypress** - Cypress engine
+
+### TAO definitions
+
+Test data access object. It is similar to DAO for test purposes <https://en.wikipedia.org/wiki/Data_access_object>
+
+### Browser object
+
+Intermediate level between widgets and test framework (cypress in this case).
+It is available as Cypress window property (`window.browser`) for step definitions and as `this.browser` for any widget and tao.
+Any cypress functions or commands should be used only through the browser object.
+There are 4 types of methods in the browser object:
+
+1. actions
+2. assertions
+3. getters
+4. utils
```

Comment:
Maybe let's give a description/definition for each of the method type?

Reply:
Rewrote this with a better explanation